### PR TITLE
mastodon: seperate localDomain from hosting and add webDomain cfg

### PR DESCRIPTION
**Pull request by ghuntley**
###### Description of changes

Opening up early for discussion -> please note that I have **NOT** tested this change.

- I launched `fediverse.ghuntley.com` as `@ghuntley@fediverse.ghuntley.com` but it looked ugly.
- Upon studying [the documentation](https://docs.joinmastodon.org/admin/config/) I learned that _it is indeed_ possible to do `@ghuntley@ghuntley.com` if I configure a webfinger 301.
- Currently localDomain and web-serving is coupled in nixpkgs leading people to make mistakes similar to the one I made. 
- I am proposing that we remove this coupling to ensure others don't make the same mistake again.

Related https://twitter.com/GeoffreyHuntley/status/1591026513629319169

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/mastodon.nix [hunk 1]
```diff
@@ -637,7 +644,7 @@ in {
     services.nginx = lib.mkIf cfg.configureNginx {
       enable = true;
       recommendedProxySettings = true; # required for redirections to work
-      virtualHosts."${cfg.localDomain}" = {
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:640]: Some form of migration is required for existing users. What and how?

Comment (ncfavier) [reply nixos/modules/services/web-apps/mastodon.nix:640]: Default `webDomain` to `localDomain`?

File: nixos/modules/services/web-apps/mastodon.nix [hunk 2]
```diff
@@ -213,6 +214,12 @@ in {
 
       localDomain = lib.mkOption {
         description = lib.mdDoc "The domain serving your Mastodon instance.";
+        example = "example.org";
+        type = lib.types.str;
+      };
+
+      webDomain = lib.mkOption {
+        description = lib.mdDoc "The web domain serving your Mastodon instance.";
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:222]:
```suggestion
        description = lib.mdDoc "The web domain serving your Mastodon instance. If different to localDomain then a web-finger 301 redirect must be configured. See https://docs.joinmastodon.org/admin/config/";
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 3]
```diff
@@ -213,6 +214,12 @@ in {
 
       localDomain = lib.mkOption {
         description = lib.mdDoc "The domain serving your Mastodon instance.";
+        example = "example.org";
+        type = lib.types.str;
+      };
+
+      webDomain = lib.mkOption {
+        description = lib.mdDoc "The web domain serving your Mastodon instance. If different to localDomain then a web-finger 301 redirect must be configured. See https://docs.joinmastodon.org/admin/config/";
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:222]:
```suggestion
        description = lib.mdDoc "The web domain serving your Mastodon instance. If different to localDomain then a manual web-finger 301 redirect must be configured. See https://docs.joinmastodon.org/admin/config/";
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 4]
```diff
@@ -213,6 +214,12 @@ in {
 
       localDomain = lib.mkOption {
         description = lib.mdDoc "The domain serving your Mastodon instance.";
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:216]:
```suggestion
        description = lib.mdDoc "The domain serving your Mastodon instance. Configure with example.com here to enable pretty @username@example.com style usernames";
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 5]
```diff
@@ -212,7 +213,13 @@ in {
       };
 
       localDomain = lib.mkOption {
-        description = lib.mdDoc "The domain serving your Mastodon instance.";
+        description = lib.mdDoc "The domain serving your Mastodon instance. Configure with example.com here to enable pretty @username@example.com style usernames";
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:216]:
```suggestion
        description = lib.mdDoc "The domain serving your Mastodon instance. If webDomain is configured to host the application on a subdomain then enable pretty @username@example.com style usernames on your domain root is possible";
```

**Review by erictapen — COMMENTED** [review]
I like this feature. Could you format the option description like in the rest of the module, so they are not super long lines?

File: nixos/modules/services/web-apps/mastodon.nix [hunk 6]
```diff
@@ -212,7 +213,13 @@ in {
...
-        description = lib.mdDoc "The domain serving your Mastodon instance.";
+        description = lib.mdDoc "The domain serving your Mastodon instance. If webDomain is configured to host the application on a subdomain then enable pretty @username@example.com style usernames on your domain root is possible";
+        example = "example.org";
+        type = lib.types.str;
+      };
+
+      webDomain = lib.mkOption {
+        description = lib.mdDoc "The web domain serving your Mastodon instance. If different to localDomain then a manual web-finger 301 redirect must be configured. See https://docs.joinmastodon.org/admin/config/";
         example = "social.example.org";
         type = lib.types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:224]:
```suggestion
        type = lib.types.str;
        default = cfg.localDomain;
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 7]
```diff
@@ -212,9 +213,27 @@ in {
...
+        '';
+        example = "example.org";
+        type = lib.types.str;
+      };
+
+      webDomain = lib.mkOption {
+        description = lib.mdDoc ''
+          The web domain serving your Mastodon instance. If different to
+          localDomain then a manual web-finger 301 redirect must be
+          configured. See https://docs.joinmastodon.org/admin/config/
```
Comment (ghuntley) [inline nixos/modules/services/web-apps/mastodon.nix:232]:
```suggestion
          configured. Study https://docs.joinmastodon.org/admin/config/ before configuring
          as this value cannot be changed once other servers federate
          with you.
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 8]
```diff
@@ -212,9 +213,29 @@ in {
...
+        description = lib.mdDoc ''
+          The web domain serving your Mastodon instance. If different to
+          localDomain then a manual web-finger 301 redirect must be
+          configured. Study https://docs.joinmastodon.org/admin/config/ 
+          before configuring as this value cannot be changed once other
+          servers federate with you.
+        '';
         example = "social.example.org";
         type = lib.types.str;
+        default = cfg.localDomain;
```
Comment (ncfavier) [inline nixos/modules/services/web-apps/mastodon.nix:238]:
```suggestion
        default = cfg.localDomain;
        defaultText = lib.literalexpression "localDomain";
```

**Review by ncfavier — CHANGES_REQUESTED** [review]
Please address the editorconfig failure (remove trailing whitespace).

**Comment by ncfavier** [comment]
We should probably take care of setting up the 301 redirection ourselves if `configureNginx` is enabled and `localDomain != webDomain`. This assumes the web domain and local domain are on the same machine, which is probably a common enough setup? We could also make it conditional on an additional `configureNginxRedirections` option.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 9]
```diff
@@ -212,9 +213,30 @@ in {
...
+        defaultText = lib.literalexpression "localDomain";
```
Comment (ncfavier) [inline nixos/modules/services/web-apps/mastodon.nix:239]:
```suggestion
        defaultText = lib.literalExpression "localDomain";
```

Sorry, typo'd

File: nixos/modules/services/web-apps/mastodon.nix [hunk 10]
```diff
@@ -212,9 +213,30 @@ in {
       };
 
       localDomain = lib.mkOption {
-        description = lib.mdDoc "The domain serving your Mastodon instance.";
+        description = lib.mdDoc ''
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:216]: After reading both documentation strings, it is still not clear to me from first glance what is what.

I think the main problem I had comes from both first sentences being very similar. How about "The domain that is going to show up in usernames" and "The domain serving the Mastodon web interface." as first sentences?

**Comment by erictapen** [comment]
> We should probably take care of setting up the 301 redirection ourselves if `configureNginx` is enabled and `localDomain != webDomain`. This assumes the web domain and local domain are on the same machine, which is probably a common enough setup? We could also make it conditional on an additional `configureNginxRedirections` option.

This hasn't been implemented yet, right? I'm unsure how I feel about this. Imo this more belongs into the Mastodon wiki, as the assumption that both domains are served from the same host is not checkable at evaluation time and `configureNginxRedirections` would add complexity with comparatively little benefits.

I'm just hesitant to grow the scope of the module too much, especially when it comes to new user facing knobs, that need to be tested and maintained in the future.

**Comment by SuperSandro2000** [comment]
Closing as stale and as there was hesitation from the maintainer.

**Pull request closed without merging**